cachi2/core/package_managers/yarn_classic/main.py

```diff
@@ -16,3 +81,27 @@ def _generate_build_environment_variables() -> list[EnvironmentVariable]:
     }
 
     return [EnvironmentVariable(name=key, value=value) for key, value in env_vars.items()]
+
+
+def _verify_corepack_yarn_version(source_dir: RootedPath, env: dict[str, str]) -> None:
```

Comment:
I'd like to have one shared function for both yarn modules to check the Yarn version and ensure it is within some range of acceptable values. I started experimenting with using [semantic_version](https://github.com/rbarrois/python-semanticversion) instead of semver since they have a nice [SimpleSpec](https://github.com/rbarrois/python-semanticversion#using-a-specification) class that does what I was looking for.

If others like that idea, I can continue with it

Reply:
> I started experimenting with using [semantic_version](https://github.com/rbarrois/python-semanticversion) instead of semver since they have a nice [SimpleSpec](https://github.com/rbarrois/python-semanticversion#using-a-specification) class that does what I was looking for.

The project hasn't made a release in 2 years, seems like it has a single maintainer and when I look at PRs, issues, not much has been done really in the past year. For what it seems you need to do, pulling in another depedency seems like an overkill.
IIUC you want to be able to do:
```
Version("1.2.3") in SimpleSpec(">=1.0.6", "<4")
```

I'm not a fan of depending on the semver library in general for what we actually need to check, given the python's packaging version handling seems to cover everything semver does and some more (Yes, I know, it's specific to Python packaging only, but admittedly our use case touches only the surface). So, I would propose having a look at https://packaging.pypa.io/en/stable/specifiers.html#packaging.specifiers.SpecifierSet.__contains__  which is a generic class that seems to be doing the same thing as `SimpleSpec` and supports standard Python comparison operators, have a look at the examples if you think it covers your use case.

Reply:
I actually started off using packaging, but cut that work short due to the fact that it [didn't appear to be strictly compatible with semver](https://packaging.python.org/en/latest/specifications/version-specifiers/#semantic-versioning) and the [version field in package.json can be any valid semver](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#version). I can take another look at it and see if it can be enough for what we need though :+1:

Reply:
> I actually started off using packaging, but cut that work short due to the fact that it [didn't appear to be strictly compatible with semver](https://packaging.python.org/en/latest/specifications/version-specifiers/#semantic-versioning) and the [version field in package.json can be any valid semver](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#version).

We shouldn't depend on pre-release and local build deps,  especially for pkg mgrs that should take care of the differences, right? We're not intending to represent valid user deps with what I'm suggesting, so I think we should be fine.

Reply:
I might not quite understand what you mean with the last comment. Do you mean that we should ignore the prerelease and build parts of the semver? They're definitely edge cases, but we currently do not have a restriction on the user specifying an rc for Yarn v2+.

I think the packaging SpecifierSet would work well for the comparison, but I think the parsing should be done with semver, since the input is a semver. It should be possible to translate one to the other fairly easily.

Reply:
> They're definitely edge cases, but we currently do not have a restriction on the user specifying an rc for Yarn v2+

I totally neglected this aspect, I'm sorry.

> I think the packaging SpecifierSet would work well for the comparison, but I think the parsing should be done with semver, since the input is a semver. It should be possible to translate one to the other fairly easily.

It's the pre-release and dev releases that need to be translated though, correct?

Reply:
Yeah, the prerelease part. I think we might be able to ignore the build part since according to the semver spec it should not be used for determining precedence.

The tricky thing appears to be that packaging is a lot less permissive when it comes to the contents of the prerelease part. Valid semver prereleases don't equal valid packaging prereleases.

Reply:
> The tricky thing appears to be that packaging is a lot less permissive when it comes to the contents of the prerelease part. Valid semver prereleases don't equal valid packaging prereleases.

Yes, I've seen the examples. But I wonder how it impacts us. Even if users are allowed to specify Yarn prerelease versions, all we care about are the final releases isn't it?
So if we define the spec as `(">=1.6.2, <4")` then what is preventing us to just check the versions simply based on  `major.minor.micro` and iff the input prerelease semver falls into the range then check the respective range boundaries for equality with the input semver (without prerelease) and if matched, deny it. Otherwise it means that the semver string (including prereleases) is well within the range and hence we don't care. In that case we don't even need to translate anything.

Reply:
Filed https://github.com/containerbuildsystem/cachi2/issues/660 for this